#  Descriptive Statistics
### Learning Objectives
* Compute mean, median, and mode and choose the appropriate one for a given dataset.
* Compute and interpret variance, standard deviation, and IQR.
* Explain how outliers affect each measure differently.
### Key Topics
* Why descriptive statistics precedes modeling
* Central tendency: mean, median, mode
* Spread: range, variance, standard deviation, IQR
* Percentiles and quartiles

### 1.1 Statistic befor modeling
A machine learning model is basically a summary of patterns in data. To judge if a model makes sense, you first need to understand the dataset — its average, variation, and overall shape. Descriptive statistics give you the precise language to do that, forming the foundation for exploring and evaluating data.

### 1.2  Measures of Central Tendency

* Mean: The arithmetic average ,Symmetric data with no extreme outliers
* Median: The middle value when sorted, Skewed data or data with outliers (more robust)
* Mode: The most frequent value, Categorical data, or finding the most common value

In [8]:
import numpy as np
data =np.array([5,7,4,2,6,9,5])
print("mean: ",np.mean(data))
print("median: ",np.median(data))


mean:  5.428571428571429
median:  5.0


In [9]:
import scipy.stats as stats

print("mode: ",stats.mode(data))

mode:  ModeResult(mode=np.int64(5), count=np.int64(2))


### 1.3  Measures of Spread
Spread shows how far data points lie from the average, Two datasets can have the same mean but look very different if one is tightly clustered and the other widely scattered.


* Range: Maximum minus minimum — simple but very sensitive to outliers
* Variance: The average squared distance from the mean
* Standard deviation: The square root of variance — spread in the same units as the data
* IQR (interquartile range): The range of the middle 50% of the data (Q3 - Q1), robust to outliers

In [10]:



data2= np.array([4,8,9,3,5,6,7])
print("max: ",data.max())
print("min: ",data.min())
print("range: ", data.max()-data.min())
print("Standard deviation: ", np.std(data))
print("variance: ", np.var(data))
q1, q3 = np.percentile(data, [25, 75])
iqr = q3 - q1
print("IQR:", iqr)



max:  9
min:  2
range:  7
Standard deviation:  2.0603150145508513
variance:  4.244897959183674
IQR: 2.0


rs. Box plots display these values, making them key for spotting outliers.### 1.4  Percentiles and Quartiles

Percentiles show the cutoff where a certain share of data lies below. The 25th (Q1), 50th (median), and 75th (Q3) split data into quarte

In [11]:
# computing everything at once
import pandas as pd
ages = pd.Series([22, 24, 21, 23, 25, 22, 26, 24, 21, 23, 45])   # 45 is an intentional outlier

print(ages.describe())

count    11.000000
mean     25.090909
std       6.789029
min      21.000000
25%      22.000000
50%      23.000000
75%      24.500000
max      45.000000
dtype: float64


### Hands-On Lab: Describing a Real Dataset
* Step 1: Load a numeric column from a provided dataset into a NumPy array or Pandas Series.

In [12]:
import pandas as pd
import numpy as np
import scipy.stats as stats

In [13]:
df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [14]:
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [15]:
df=df.dropna()

In [16]:
df.isnull().sum()


survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

In [17]:
print(df.shape)

(182, 15)


* Step 2: Compute its mean, median, mode, standard deviation, and IQR.

In [18]:
mean= np.mean(df["fare"])
median=np.median(df["fare"])
mode =stats.mode(df["fare"])
stde=np.std(df["fare"])
q1, q3 = np.percentile(df["fare"], [25, 75])
iqr = q3 - q1
print("mean: ",mean)
print("median: ",median)
print("mode: ",mode)
print("standard deviation: ",stde)
print("IQR:", iqr)


mean:  78.91973516483515
median:  57.0
mode:  ModeResult(mode=np.float64(26.55), count=np.int64(7))
standard deviation:  76.28034507367532
IQR: 60.3


* Step 3: Identify whether the mean or median better represents a 'typical' value, and justify the choice.

The mean fare ($78.92) is higher than the median ($57.00), showing right‑skew. A few high fares raise the mean, while most paid closer to the median.


* Step 4: Write a Markdown cell summarizing the column's center and spread in plain language. 

The fare column has a mean of $78.92 and a median of $57.00, showing a right‑skewed distribution. A small group of passengers with very high fares pulls the average upward, while most paid closer to the median. The mode ($26.55, appearing 7 times) highlights a cluster of low fares, reinforcing that the majority spent less. The large standard deviation ($76.28) and wide IQR ($60.30) confirm the data is spread out, with extreme high fares driving much of that variation. Altogether, this makes the median the most reliable measure of a typical fare, since it isn’t distorted by outliers.